### PR TITLE
[android][image-picker] Use a more modern and streamlined system image picker

### PR DESCRIPTION
**Pull request by fobos531**
# Why

See https://developer.android.com/training/data-storage/shared/photopicker

This brings the usage and the UX of the image picker more in-line with what we have on iOS - the interface is very, very similar. Additionally, it brings more feature parity with iOS - before, we didn't limit the number of photos/videos the user can select on Android, and with this picker its now possible (I assume it was possible before as well, it was just not done for some reason).

Additionally, this photo picker is backwards compatible all the way back to KitKat - see https://android-developers.googleblog.com/2023/04/photo-picker-everywhere.html

# How

Upgraded a couple of dependencies according to Android's docs, and edited the ImageLibraryContract to use the intents provided by `PickVisualMedia()` and `PickMultipleVisualMedia()`, respectively.

# Test Plan

Verified that all tests still pass and that I can select images & videos using the available options.

**Note**: As mentioned before, I bumped a couple of dependencies, as required by Android's documentation for the backported image picker functionality, most importantly `androidx.activity` and `androidx.fragment`. I'm not entirely sure of the negative implications of bumping these dependencies, so let me know if it's a no-go.

I understand this might be a pretty significant change and that it might be better to discard it for now for compatibility purposes, but I will be happy nevertheless if it ends up getting merged haha

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt [hunk 1]
```diff
@@ -22,6 +22,10 @@ internal class ImagePickerOptions : Record, Serializable {
   @FloatRange(from = 0.0, to = 1.0)
   var quality: Double = 0.2
 
+  @Field
+  @IntRange(from = 0)
+  var selectionLimit: Int = 0
```
Comment (tsapeta) [inline packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:27]: If `0` indicates no limit, could you please add a comment here or make it like on iOS where it's set to `UNLIMITED_SELECTION` constant?https://github.com/expo/expo/blob/887aceac3db640a2900961ccff84eebe63e718c3/packages/expo-image-picker/ios/ImagePickerOptions.swift#L54

File: packages/expo-modules-core/CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,8 @@
 
 ### 💡 Others
 
+- Updated `androidx.activity:activity-ktx` to `1.7.1` and `androidx.fragment:fragment-ktx` to `1.7.1` [#22658](https://github.com/expo/expo/pull/22658) by [@fobos531](https://github.com/fobos531)
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:23]:
```suggestion
- Updated `androidx.activity:activity-ktx` to `1.7.1` and `androidx.fragment:fragment-ktx` to `1.5.7` [#22658](https://github.com/expo/expo/pull/22658) by [@fobos531](https://github.com/fobos531)
```

File: docs/public/static/data/v48.0.0/expo-image-picker.json [hunk 3]
```diff
@@ -1 +1 @@
...
...est permissions to access the media library.\nThis uses both "},{"kind":"code","text":"`requestMediaLibraryPermissionsAsync`"},{"kind":"text","text":" and "},{"kind":"code","text":"`getMediaLibraryPermissionsAsync`"},{"kind":"text","text":" to interact with the permissions."}],"blockTags":[{"tag":"@example","content":[{"kind":"code","text":"```ts\nconst [status, requestPermission] = ImagePicker.useMediaLibraryPermissions();\n```"}]}]},"parameters":[{"name":"options","kind":32768,"flags":{"isOptional":true},"type":{"type":"reference","typeArguments":[{"type":"reflection","declaration":{"name":"__type","kind":65536,"children":[{"name":"writeOnly","kind":1024,"flags":{"isOptional":true},"type":{"type":"intrinsic","name":"boolean"}}]}}],"name":"PermissionHookOptions"}}],"type":{"type":"tuple","elements":[{"type":"union","types":[{"type":"literal","value":null},{"type":"reference","name":"MediaLibraryPermissionResponse"}]},{"type":"reference","typeArguments":[{"type":"reference","name":"MediaLibraryPermissionResponse"}],"name":"RequestPermissionMethod"},{"type":"reference","typeArguments":[{"type":"reference","name":"MediaLibraryPermissionResponse"}],"name":"GetPermissionMethod"}]}}]}]}
```
Comment (tsapeta) [inline docs/public/static/data/v48.0.0/expo-image-picker.json:1]: Can you revert docs data for SDK 48? We will not backport this change to SDK 48 😉

**Review by tsapeta — CHANGES_REQUESTED** [review]
Nice, thanks for implementing that! People will love it 😍 Left a few comments but generally looks good. I quickly checked changes in these AndroidX dependencies, but looks like they shouldn't cause any regressions.

**Review by tsapeta — APPROVED** [review]
Great, thanks! Some CI checks are failing but they don't seem to be related to this PR, so I'm merging it.

**Comment by thespacemanatee** [comment]
I'm experiencing a fresh issue on crashlytics due to the bumping of dependencies from this PR. Will add more info in due time but here is the event:

Screenshot 2023-06-22 at 12 42 34 AM

**Comment by younes0** [comment]
@fobos531 Unfortunately, there's no more [_Browse options_](https://media.discordapp.net/attachments/1186164915984138322/1186164916193857626/image.png?ex=65c065a9&is=65adf0a9&hm=f514b04b69c5fb564a625f8469ef7dc2ef7ab27286c8e93fdeb0946a5fc77ba9&=&format=webp&quality=lossless) on Android with these changes.
Browse options allows User to select pictures from Google Drive, Photos & others.
See: https://stackoverflow.com/a/77317553/822837

image

**Comment by wodin** [comment]
I also had to patch `expo-image-picker` for my app. Unfortunately, you can't pick multiple images:

```diff
diff --git a/node_modules/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt b/node_modules/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt
index ff15c91..09f94b0 100644
--- a/node_modules/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt
+++ b/node_modules/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt
@@ -5,9 +5,8 @@ import android.content.ContentResolver
 import android.content.Context
 import android.content.Intent
 import android.net.Uri
-import androidx.activity.result.PickVisualMediaRequest
-import androidx.activity.result.contract.ActivityResultContracts.PickVisualMedia
-import androidx.activity.result.contract.ActivityResultContracts.PickMultipleVisualMedia
+import android.util.Log
+import androidx.activity.result.contract.ActivityResultContracts.GetContent
 import expo.modules.imagepicker.ImagePickerOptions
 import expo.modules.imagepicker.MediaTypes
 import expo.modules.imagepicker.UNLIMITED_SELECTION
@@ -33,70 +32,29 @@ internal class ImageLibraryContract(
       ?: throw Exceptions.ReactContextLost()
 
   override fun createIntent(context: Context, input: ImageLibraryContractOptions): Intent {
-    val request = PickVisualMediaRequest.Builder()
-      .setMediaType(
-        when (input.options.mediaTypes) {
-          MediaTypes.VIDEOS -> {
-            PickVisualMedia.VideoOnly
-          }
+    Log.d("ImagePicker", "Creating intent for image selection")
+    val options = input.options
 
-          MediaTypes.IMAGES -> {
-            PickVisualMedia.ImageOnly
-          }
-
-          else -> {
-            PickVisualMedia.ImageAndVideo
-          }
-        }
-      )
-      .build()
-
-    if (input.options.allowsMultipleSelection) {
-      val selectionLimit = input.options.selectionLimit
-
-      if (selectionLimit == 1) {
-        // If multiple selection is allowed but the limit is 1, we should ignore
-        // the multiple selection flag and just treat it as a single selection.
-        return PickVisualMedia().createIntent(context, request)
-      }
-
-      if (selectionLimit > 1) {
-        return PickMultipleVisualMedia(selectionLimit).createIntent(context, request)
-      }
-
-      // If the selection limit is 0, it is the same as unlimited selection.
-      if (selectionLimit == UNLIMITED_SELECTION) {
-        return PickMultipleVisualMedia().createIntent(context, request)
-      }
+    // If multiple selection is not allowed or the selection limit is 1, use GetContent
+    if (!options.allowsMultipleSelection || options.selectionLimit == 1) {
+      return GetContent().createIntent(context, "image/*")
+    } else {
+      // Handle multiple selection scenario
+      // Since GetContent does not support multiple selections, you need to implement a custom solution
+      // or use a different approach that supports multiple selections.
+      throw UnsupportedOperationException("Multiple selection is not supported with GetContent")
     }
-
-    return PickVisualMedia().createIntent(context, request)
   }
 
   override fun parseResult(input: ImageLibraryContractOptions, resultCode: Int, intent: Intent?) =
     if (resultCode == Activity.RESULT_CANCELED) {
+      Log.d("ImagePicker", "Image selection cancelled")
       ImagePickerContractResult.Cancelled
     } else {
-      intent?.takeIf { resultCode == Activity.RESULT_OK }?.getAllDataUris()?.let { uris ->
-        if (input.options.allowsMultipleSelection) {
-          ImagePickerContractResult.Success(
-            uris.map { uri ->
-              uri.toMediaType(contentResolver) to uri
-            }
-          )
-        } else {
-          if (intent.data != null) {
-            intent.data?.let { uri ->
-              val type = uri.toMediaType(contentResolver)
-              ImagePickerContractResult.Success(listOf(type to uri))
-            }
-          } else {
-            uris.firstOrNull()?.let { uri ->
-              val type = uri.toMediaType(contentResolver)
-              ImagePickerContractResult.Success(listOf(type to uri))
-            } ?: ImagePickerContractResult.Error
-          }
-        }
+      Log.d("ImagePicker", "Processing image selection result")
+      intent?.data?.let { uri ->
+        val type = uri.toMediaType(contentResolver)
+        ImagePickerContractResult.Success(listOf(type to uri))
       } ?: ImagePickerContractResult.Error
     }
 }
```

Also, a couple of days ago I happened to see another patch for this problem here that I've been meaning to try out:
https://github.com/mozzius/graysky/blob/main/patches/expo-image-picker%4014.7.1.patch

**Pull request merged**